src/Chapter3.hs

```diff
@@ -1005,8 +1178,10 @@ properties using typeclasses, but they are different data types in the end.
 Implement data types and typeclasses, describing such a battle between two
 contestants, and write a function that decides the outcome of a fight!
 -}
-
-
+-- sorry, a bit too foggy for me :-( having hard time translating this "knight and monster domain" into typeclasses.
+-- doubts: 1) knights and monsters do their activities *in turns* but then each fighter has a list of actions
+-- to be performed *in sequence* and then again from the beginning...it seems a bit contradicting to me.
+-- 2) what's the meaning of action like "drink Health potion" or "cast a spell"?
```

Comment:
I can give a few hints, so you can crack the advanced task 🙂 

For this task you need to create two data types: `Monster` and `Knight`. Each type will have the properties described in the task description.

In addition, I would recommend creating separate types for `MonsterAction` and `KnightAction` to represent possible actions for each type correspondingly.

Regarding the list of actions, let's look at the example. A `Knight` can have initially 3 actions:

`Attack, DrinkPotion, Attack`

So, if the `Knight` survives 8 turns, their actions will be: 

`Attack, DrinkPotion, Attack, Attack, DrinkPotion, Attack, Attack, DrinkPotion`

`DrinkHealthPotion` should increase the health of a `Knight` and `Cast a spell` should increase defence of a knight, as written in the task description: `cast a spell to increase their defence`

Since each data type (`Monster` and `Knight`) can be fighters, I can suggest creating a typeclass of the following shape:

```haskell
class Fighter a where
    ... you need to think of possible methods here ...
```

So you can them implement a recursive function using this typeclass:

```haskell

data FightResult a b
    = FirstWins a
    | SecondWins b

fight :: (Fighter a, Fighter b) => a -> b -> FightResult a b
```

I hope this clarifies the task a bit more and gives some ideas on how to proceed with it :slightly_smiling_face: